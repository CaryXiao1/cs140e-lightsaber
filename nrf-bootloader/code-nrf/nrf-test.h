#ifndef __NRF_TEST_H__
#define __NRF_TEST_H__
// various support routines for testing.  
#include "rpi.h"
#include "nrf.h"

// left side NRF pin-out of parthiv's board (server)
static inline nrf_conf_t
parthiv_left(uint32_t nbytes, uint32_t ch) {
    nrf_conf_t c = nrf_conf_mk(nbytes, ch);
    c.spi_chip = 0;
    c.ce_pin = 6;
    c.int_pin = 22;
    return c;
}

// right side NRF pin-out of parthiv's board (client)
static inline nrf_conf_t
parthiv_right(uint32_t nbytes, uint32_t ch) {
    nrf_conf_t c = nrf_conf_mk(nbytes, ch);
    c.spi_chip = 1;
    c.ce_pin = 5;
    c.int_pin = 23;
    return c;
}

static inline nrf_conf_t
client_conf(uint32_t nbytes) {
    return parthiv_right(nbytes,nrf_default_channel);
}
static inline nrf_conf_t
server_conf(uint32_t nbytes) {
    return parthiv_left(nbytes,nrf_default_channel);
}

// server wrappers for making an nrf: <ack_p> indicates ack'd or not.
static inline nrf_t *
server_mk(uint32_t rxaddr, uint32_t nbytes, int ack_p) {
    return nrf_init(server_conf(nbytes), rxaddr, ack_p);
}
static inline nrf_t *
server_mk_ack(uint32_t rxaddr, uint32_t nbytes) {
    return server_mk(rxaddr, nbytes, 1);
}
static inline nrf_t *
server_mk_noack(uint32_t rxaddr, uint32_t nbytes) {
    return server_mk(rxaddr, nbytes, 0);
}

// client wrappers.
static inline nrf_t *
client_mk(uint32_t rxaddr, uint32_t nbytes, int ack_p) {
    return nrf_init(client_conf(nbytes), rxaddr, ack_p);
}
static inline nrf_t *
client_mk_ack(uint32_t rxaddr, uint32_t nbytes) {
    return client_mk(rxaddr, nbytes, 1);
}
static inline nrf_t *
client_mk_noack(uint32_t rxaddr, uint32_t nbytes) {
    return client_mk(rxaddr, nbytes, 0);
}
#endif
