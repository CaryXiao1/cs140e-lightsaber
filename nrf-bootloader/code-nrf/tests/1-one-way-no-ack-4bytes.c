// one way test of 4-byte not-ack'd packets
#include "nrf-test.h"

// useful to mess around with these. 
enum { ntrial = 1000, timeout_usec = 1000 };

// for simplicity the code is hardwired for 4 byte packets (the size
// of the data we send/recv).
static void
one_way_noack(nrf_t *server, nrf_t *client, int verbose_p) {
    unsigned client_addr = client->rxaddr;
    unsigned ntimeout = 0, npackets = 0;

    // keep sending the loop variable to ourselves, and make sure we receive
    // it.
    for(unsigned i = 0; i < ntrial; i++) {
        // so we can see what kind of progress is being made.
        if(verbose_p && i  && i % 100 == 0)
            trace("sent %d no-ack'd packets\n", i);

        // send from server to client.
        nrf_send_noack(server, client_addr, &i, 4);

        // receive on client nic
        uint32_t x;
        if(nrf_read_exact_timeout(client, &x, 4, timeout_usec) == 4) {
            // we aren't doing acks, so can easily lose packets.  [i.e.,
            // it's not actually an error in the code.]
            if(x != i) {
                nrf_output("lost/dup packet: received %d (expected=%d)\n", x,i);
                client->tot_lost++;
            }
            npackets++;
        } else {
            if(verbose_p) 
                nrf_output("receive failed for packet=%d\n", i);
            ntimeout++;
        }
    }
    trace("trial: successfully sent %d no-ack'd pkts, [lost=%d, timeouts=%d]\n",
        npackets, client->tot_lost, ntimeout, ntimeout);
}

void notmain(void) {
    unsigned nbytes = 4;

    trace("configuring no-ack server=[%x] with %d nbyte msgs\n",
                server_addr, nbytes);
    // nrf-test.h
    nrf_t *s = server_mk_noack(server_addr, nbytes);
    nrf_dump("unreliable server config:\n", s);

    trace("configuring no-ack client=[%x] with %d nbyte msg\n",
                client_addr, nbytes);
    // nrf-test.h
    nrf_t *c = client_mk_noack(client_addr, nbytes);
    nrf_dump("unreliable client config:\n", c);


    // reset the times so we get a bit better measurement.
    nrf_stat_start(s);
    nrf_stat_start(c);

    // do the test
    one_way_noack(s, c, 1);

    // print the stats.
    nrf_stat_print(s, "server: done with one-way test");
    nrf_stat_print(c, "client: done with one-way test");
}
