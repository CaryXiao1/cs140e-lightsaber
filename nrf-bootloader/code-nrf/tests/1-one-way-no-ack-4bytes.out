TRACE:notmain:configuring no-ack server= with 4 nbyte msgs
TRACE:notmain:configuring no-ack client= with 4 nbyte msg
TRACE:one_way_noack:sent 100 no-ack'd packets
TRACE:one_way_noack:sent 200 no-ack'd packets
TRACE:one_way_noack:sent 300 no-ack'd packets
TRACE:one_way_noack:sent 400 no-ack'd packets
TRACE:one_way_noack:sent 500 no-ack'd packets
TRACE:one_way_noack:sent 600 no-ack'd packets
TRACE:one_way_noack:sent 700 no-ack'd packets
TRACE:one_way_noack:sent 800 no-ack'd packets
TRACE:one_way_noack:sent 900 no-ack'd packets
TRACE:one_way_noack:trial: successfully sent 1000 no-ack'd pkts, 
